In [ ]:
%%bash
rm -fv ipopt*
apt install -q -yy libgfortran5
sleep 1
wget -q https://www.ieor.iitb.ac.in/files/faculty/amahajan/ipopt
wget -q https://www.ieor.iitb.ac.in/files/faculty/amahajan/test.nl
chmod u+rx ipopt

removed 'ipopt'
Reading package lists...
Building dependency tree...
Reading state information...
libgfortran5 is already the newest version (8.4.0-1ubuntu1~18.04).
0 upgraded, 0 newly installed, 0 to remove and 21 not upgraded.


In [ ]:
!./ipopt test.nl



******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt
******************************************************************************

This is Ipopt version 3.13.2, running with linear solver mumps.
NOTE: Other linear solvers might be more efficient (see Ipopt documentation).

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:      230

Total number of variables............................:       30
                     variables with only lower bounds:       30
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equ

In [ ]:
!pip install -q pyomo
import numpy as np
from pyomo.environ import *
import pandas as pd

**1. ANSWER:** \
**We have the following optimization problem from lab6 :**

 \begin{array}{l}
Here\ b_{0} \ ,\ b_{1} \ and\ b_{2} \ are\ the\ decision\ variables\ as\ described\ in\ the\ question.\\
\bf{Min}\\
\> \> \> \ \ \ \ \ \  \ \ \sum ^{6}_{i=1} |R_{i} | \\
where\  \ R_{i} =\ P_{i} \ -\widetilde{P_{i}}\\
and\ \ \ \ \ \tilde{P}_{i} \ =b_{0} +b_{1} L_{i} +b_{2} E_{i} \ \ \ \ \ ;\ \ i=\ 1,2,\ .\ .\ .\ ,6.\\
\\
s.t\\
b_{0}\geqslant 0 \ , \ and \ b_{1},\ b_{2}\ \epsilon\ \bf{R}\ \\
\end{array}

**Now, Problem solving Approach for lab7 is as follows:** 

 \begin{array}{l}
Let\ \ \ \ \ \ \ \ \ \ e_{i} \ =\ R_{i} \\
Then\ the\ objective\ funtion\ reduces \ to\ :\\
\\
\bf{Min}\ \ \ \ \ \ \sum ^{6}_{i=1} e_{i}^2 \\
b_{0}\geqslant 0 \ , \ and \ b_{1},\ b_{2}\ \epsilon\ \bf{R}
\end{array}

In [ ]:
model = ConcreteModel()
M=[0,1,2]
model.b = Var(M)
model.b[0].setlb(0)


model.obj = Objective(expr = (175000-(model.b[0]+1800*model.b[1]+250*model.b[2]))**2+(150000-(model.b[0]+2800*model.b[1]+400*model.b[2]))**2+
                             (125000-(model.b[0]+3500*model.b[1]+450*model.b[2]))**2+(90000-(model.b[0]+800*model.b[1]+100*model.b[2]))**2+
                              (80000-(model.b[0]+700*model.b[1]+200*model.b[2]))**2+(120000-(model.b[0]+500*model.b[1]+50*model.b[2]))**2, sense=minimize)

model.pprint()

1 Set Declarations
    b_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    b : Size=3, Index=b_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True :  Reals
          1 :  None :  None :  None : False :  True :  Reals
          2 :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : (175000 - (b[0] + 1800*b[1] + 250*b[2]))**2 + (150000 - (b[0] + 2800*b[1] + 400*b[2]))**2 + (125000 - (b[0] + 3500*b[1] + 450*b[2]))**2 + (90000 - (b[0] + 800*b[1] + 100*b[2]))**2 + (80000 - (b[0] + 700*b[1] + 200*b[2]))**2 + (120000 - (b[0] + 500*b[1] + 50*b[2]))**2

3 Declarations: b_index b obj


In [ ]:
opt_ipopt = SolverFactory('ipopt')
result = opt_ipopt.solve(model)
print('Status :',result.solver.status)
print('Termination Condition :',result.solver.termination_condition)
print('\nThe optimal value of the objective function is :',model.obj())
print('\nThe value of the decision variables are : ')
print()
for i in range(3):
  print('b[',i,'] = ', model.b[i].value)

Status : ok
Termination Condition : optimal

The optimal value of the objective function is : 4291134028.8074555

The value of the decision variables are : 

b[ 0 ] =  104904.4164371955
b[ 1 ] =  38.702075831391646
b[ 2 ] =  -193.3223893242956


**2.ANSWER:**\
The optimal value of the objective function is : 4291134028.8074555 \
The value of the decision variables are : \
**b[ 0 ] =  104904.4164371955 ,
b[ 1 ] =  38.702075831391646 ,
b[ 2 ] =  -193.3223893242956** \

in lab6 the absolute value
of the residuals was **125454.545** whereas in lab7 the value is **4291134028.8074555**, a huge difference we have because of the introduction of the squares of the residuals,which are more sensitive to any slight changes. 
